implementations/c/ockam/channel/channel_impl.c

```diff
@@ -0,0 +1,297 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <ctype.h>
+#include "ockam/syslog.h"
+#include "ockam/memory.h"
+#include "ockam/key_agreement.h"
+#include "ockam/key_agreement/xx.h"
+#include "ockam/transport.h"
+#include "ockam/io/impl.h"
+#include "ockam/channel.h"
+#include "channel_impl.h"
+#include "ockam/codec.h"
+
+ockam_memory_t* gp_ockam_channel_memory = NULL;
+
+ockam_error_t channel_read(void*, uint8_t*, size_t, size_t*);
+ockam_error_t channel_write(void*, uint8_t*, size_t);
+
+uint8_t g_encoded_text[MAX_CHANNEL_PACKET_SIZE];
+uint8_t g_cipher_text[MAX_CHANNEL_PACKET_SIZE];
+
+uint8_t* channel_encode_header(ockam_channel_t* p_ch, uint8_t* p_encoded)
+{
+  p_encoded = encode_ockam_wire(p_encoded);
+  if (NULL == p_encoded) goto exit;
+  *p_encoded++ = 0; //!! onward route not implemented
+  *p_encoded++ = 0; //!! return route not implemented
+exit:
+  return p_encoded;
+}
+
+uint8_t* channel_deocde_header(ockam_channel_t* p_ch, uint8_t* p_encoded)
+{
+  p_encoded = decode_ockam_wire(p_encoded);
+  if (NULL == p_encoded) goto exit;
```

Comment:
Use either (p_encoded) and (!p_encoded) or (NULL != p_encoded) and (NULL == p_encoded)